Add cache support to inventory plugin

p3ck opened the pull request:
##### SUMMARY
The default cache_plugin is memory which won't really help. Change to jsonfile to cache on disk.

cache: true
cache_plugin: ansible.builtin.jsonfile
cache_timeout: 7200
cache_connection: /tmp/azure_inventory

Fixes #700 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/inventory/azure_rm.py

##### ADDITIONAL INFORMATION

Even with just 3 hosts we see a time savings

```paste below
(venv-azure-3.11) [admin@okd-master-1 inventory]$ time ansible-inventory -i arc_azure_rm.yml --graph -v
No config file found; using defaults
Using inventory plugin 'ansible_collections.azure.azcollection.plugins.inventory.azure_rm' to process inventory source '/home/admin/inventory/arc_azure_rm.yml'
@all:
  |--@ungrouped:
  |  |--devtools-win11-runner
  |  |--ubuntu-24.04-amd64-azure
  |  |--w11

real	0m3.306s
user	0m2.320s
sys	0m0.181s
(venv-azure-3.11) [admin@okd-master-1 inventory]$ time ansible-inventory -i arc_azure_rm.yml --graph -v
No config file found; using defaults
Using inventory plugin 'ansible_collections.azure.azcollection.plugins.inventory.azure_rm' to process inventory source '/home/admin/inventory/arc_azure_rm.yml'
@all:
  |--@ungrouped:
  |  |--devtools-win11-runner
  |  |--ubuntu-24.04-amd64-azure
  |  |--w11

real	0m2.403s
user	0m2.217s
sys	0m0.169s

```

p3ck committed:
Add cache support to inventory plugin

The default cache_plugin is memory which won't really help. Change to
jsonfile to cache on disk.

cache: true
cache_plugin: ansible.builtin.jsonfile
cache_timeout: 7200
cache_connection: /tmp/azure_inventory

xuzhang3 commented:
@p3ck  How do we update the cache? Currently it depends only on the timeout? I tested this update and once this machine is added to the cache, the value is not updated even if the machine status is updated. For example, I start/stop the VM after it is cached. Cache should be able to be updated based on latest data.
![image](https://github.com/user-attachments/assets/7835b412-3b59-47ea-97fe-ee9de67deca3)

p3ck commented:
> @p3ck How do we update the cache? Currently it depends only on the timeout? I tested this update and once this machine is added to the cache, the value is not updated even if the machine status is updated. For example, I start/stop the VM after it is cached. Cache should be able to be updated based on latest data. ![image](https://private-user-images.githubusercontent.com/57888764/417019597-7835b412-3b59-47ea-97fe-ee9de67deca3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1NzU4MzQsIm5iZiI6MTc0MDU3NTUzNCwicGF0aCI6Ii81Nzg4ODc2NC80MTcwMTk1OTctNzgzNWI0MTItM2I1OS00N2VhLTk3ZmUtZWU5ZGU2N2RlY2EzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDEzMTIxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyMWFhZDVmNzhhOTIyN2VmYzQ1NDU1ODMyNDBjZjIyMTI4MThmNjEyYWQ5YzJlNDNmZjY4NDMwNGFmYjRkODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WiFTcrpIx6_PpCO0UkTdcggI_qb803ieSLTnECw_d0E)

Can you try adding the following to your test playbook?

```
- meta: refresh_inventory
```

That should force the cache to clear.  And if not I've done something wrong.

p3ck commented:
There is also --flush-cache which can be passed in on the anisble-playbook command

p3ck committed:
Update cache logic

Fixed a mistake with the cache logic when flushing the cache didn't
update the cache on disk.  This would mean if you ran again after a
flush you could get the old stale cache.

p3ck commented:
I did see an issue where if you flush-cache or meta refresh it wouldn't update the cache on disk.  I used the same logic that is referenced here: https://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html#inventory-cache

xuzhang3 commented:
> I did see an issue where if you flush-cache or meta refresh it wouldn't update the cache on disk. I used the same logic that is referenced here: https://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html#inventory-cache

That's the issue, the cache file never update even with `cache: refresh_inventory`

p3ck committed:
Add inventory cache integration tests

Also fixed trying to load the cache when it wasn't enabled.

p3ck commented:
> > I did see an issue where if you flush-cache or meta refresh it wouldn't update the cache on disk. I used the same logic that is referenced here: https://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html#inventory-cache
> 
> That's the issue, the cache file never update even with `cache: refresh_inventory`

This is fixed now.  I've added test cases that cover both ways to flush the cache.
